File: answers/elixir/test/aoc_helpers_test.exs
defmodule AocHelpersTest do
  use ExUnit.Case
  doctest AocHelpers

  test "writes output to answers.txt" do
    answer1 = "504"
    answer2 = "024"

    AocHelpers.write_answers!(answer1, answer2)
    actual = File.read!("answers.txt")

    expected = AocHelpers.format_answers(answer1, answer2)
    assert actual == expected |> Enum.join()
  end

  test "reads input from input.txt" do
    input = ["This \n", " \n", " is a test\n", "\n", "\n"]
    File.write!("input.txt", input |> Enum.join())

    actual = AocHelpers.read_input!() |> Enum.take_every(1)

    expected = input |> Enum.map(&String.trim_trailing/1)
    assert actual == expected
  end
end

File: answers/go/aoc_helpers_test.go
package write_aoc_answers

import (
	"os"
	"strings"
	"testing"
)

func TestWriteAnswers(t *testing.T) {
	answer1, answer2 := "532", "543"

	WriteAnswers(answer1, answer2)

	expectedAnswers := formatAnswers(answer1, answer2)
	output, err := readAnswers()
	if err != nil {
		t.Fatalf("Unable to read answers: %s", err)
	}
	if output != expectedAnswers {
		t.Fatalf("Got %s expected %s", output, expectedAnswers)
	}
}

func TestReadInput(t *testing.T) {
	inputTxt := "This is\na\n\nTest"
	os.WriteFile("input.txt", []byte(inputTxt), 0666)

	actual, err := ReadInput()

	if err != nil {
		t.Fatalf("Received an error when trying to read `input.txt`: %s", err)
	}
	if strings.Join(actual, "\n") != inputTxt {
		t.Fatalf("Got %s expected %s", actual, inputTxt)
	}
}

// Reads the output from a file as a string.
func readAnswers() (string, error) {
	output, err := os.ReadFile("answers.txt")
	if err != nil {
		return "", err
	}
	return string(output), nil
}
